### PR TITLE
feat: remove old same type passwordless connectors before posting

### PR DESCRIPTION
**Pull request by darcyYe**
## Summary

remove old same type passwordless connectors before posting (for passwordless connector, we only accept one working connector for each type, including SMS and email)

## Testing

Passed IT and UTs.

### CODE REVIEW

File: packages/core/src/routes/connector.test.ts [hunk 1]
```diff
@@ -188,6 +189,39 @@ describe('connector route', () => {
       });
       expect(response).toHaveProperty('statusCode', 422);
     });
+
+    it('should post a new record and delete old records with same connector type when add passwordless connectors', async () => {
```
Comment (wangsijie) [inline packages/core/src/routes/connector.test.ts:193]: nit: add a new record, or insert a new record

File: packages/core/src/routes/connector.ts [hunk 2]
```diff
@@ -122,11 +122,28 @@ export default function connectorRoutes<T extends AuthedRouter>(router: T) {
...
-        id: generateConnectorId(),
+        id: insertConnectorId,
         ...body,
       });
 
+      if (
+        connectorFactory.type === ConnectorType.Sms ||
+        connectorFactory.type === ConnectorType.Email
+      ) {
+        const logtoConnectors = await getLogtoConnectors();
```
Comment (wangsijie) [inline packages/core/src/routes/connector.ts:135]: add comment on that?

and thoes options can be achieved by a "delete where" query?

File: packages/core/src/routes/connector.test.ts (hunk 1 above)
Comment (simeng-li) [reply packages/core/src/routes/connector.test.ts:193]: replace?

File: packages/core/src/routes/connector.ts [hunk 3]
```diff
@@ -122,11 +122,28 @@ export default function connectorRoutes<T extends AuthedRouter>(router: T) {
...
+        const allSameTypeConnectorsIds = logtoConnectors
+          .filter((logtoConnector) => logtoConnector.type === connectorFactory.type)
+          .map((logtoConnector) => logtoConnector.dbEntry.id)
```
Comment (simeng-li) [inline packages/core/src/routes/connector.ts:138]: Looks like this mapping is redundant.

File: packages/core/src/routes/connector.ts (hunk 2 above)
Comment (darcyYe) [reply packages/core/src/routes/connector.ts:135]: Seems we can't, we do not store connector `type` in DB.

File: packages/core/src/routes/connector.ts [hunk 4]
```diff
@@ -122,11 +123,23 @@ export default function connectorRoutes<T extends AuthedRouter>(router: T) {
...
+        id: insertConnectorId,
         ...body,
       });
 
+      if (
+        connectorFactory.type === ConnectorType.Sms ||
+        connectorFactory.type === ConnectorType.Email
+      ) {
+        const logtoConnectors = await getLogtoConnectors();
+        const allSameTypeConnectorsIds = logtoConnectors
```
Comment (wangsijie) [inline packages/core/src/routes/connector.ts:137]: nit: the name looks strange

File: packages/core/src/routes/connector.ts (hunk 3 above)
Comment (darcyYe) [reply packages/core/src/routes/connector.ts]: We need to get the ids of all conflicting connectors.

File: packages/core/src/queries/connector.ts [hunk 5]
```diff
@@ -37,6 +37,18 @@ export const deleteConnectorById = async (id: string) => {
   }
 };
 
+export const deleteConnectorByIds = async (ids: string[], excludesConnectorId?: string) => {
```
[the remark this answers is not in the document]
Comment (darcyYe) [reply packages/core/src/queries/connector.ts]: Updated.

File: packages/core/src/routes/connector.ts (hunk 4 above)
Comment (darcyYe) [reply packages/core/src/routes/connector.ts]: renamed as `conflictingConnectorIds`

File: packages/core/src/routes/connector.ts [hunk 6]
```diff
@@ -122,11 +123,29 @@ export default function connectorRoutes<T extends AuthedRouter>(router: T) {
...
+        connectorFactory.type === ConnectorType.Sms ||
+        connectorFactory.type === ConnectorType.Email
+      ) {
+        const logtoConnectors = await getLogtoConnectors();
+        const conflictingConnectorIds = logtoConnectors
+          .filter((logtoConnector) => logtoConnector.type === connectorFactory.type)
+          .map((logtoConnector) => logtoConnector.dbEntry.id);
+        await deleteConnectorByIds(
+          conflictingConnectorIds.filter((id) => id !== insertConnectorId)
+        );
```
Comment (simeng-li) [inline packages/core/src/routes/connector.ts:146]:
```suggestion
        const conflictingConnectorIds = logtoConnectors
          .filter(({id, type}) => type === connectorFactory.type && id !== insertConnectorId)
          .map(({id}) => id);
        await deleteConnectorByIds(conflictingConnectorIds);
```

File: packages/core/src/queries/connector.ts [hunk 7]
```diff
@@ -37,6 +37,17 @@ export const deleteConnectorById = async (id: string) => {
   }
 };
 
+export const deleteConnectorByIds = async (ids: string[]) => {
+  const { rowCount } = await envSet.pool.query(sql`
+    delete from ${table}
+    where ${fields.id} in (${sql.join(ids, sql`, `)})
+  `);
+
+  if (rowCount < 1) {
```
Comment (wangsijie) [inline packages/core/src/queries/connector.ts:46]: `rowCount !== ids.length`

Comment (darcyYe) [reply packages/core/src/queries/connector.ts]: Updated

**Review by wangsijie — APPROVED** [review]

**Pull request merged**